### ISSUE TITLE
WASM support

### ISSUE THREAD

**Issue by benkuly** [mozilla/uniffi-rs#1424]
For full Kotlin Multiplatform support within [uniffi-kotlin-multiplatform-bindings](https://gitlab.com/trixnity/uniffi-kotlin-multiplatform-bindings) I need some sort of WASM support.

My idea was to expose `RustBuffer`, `RustCallStatus` and the scaffolding code with `wasm-bindgen`, but there are some problems. For example `wasm-bindgen` does not  support `MaybeUninit<RustBuffer>`. What do you think about this idea? I don't have enough experience with Rust to be able to assess this.

**Comment by arg0d** [comment]
Have you made any progress on this? I'm also investigating WASM JS bindings, and I'm gonna write my proposal for WASM JS implementation in this thread later this week.

**Comment by benkuly** [comment]
No more progess since opening the issue.

**Comment by mhammond** [comment]
At a very high abstract level, this makes perfect sense to me and is roughly the approach @bendk took for the Desktop Firefox JS bindings (although there we generate .webidl bindings and implementation etc). The fact these bindings could be external and the of the implementation: "No major changes are needed in uniffi-rs upstream repository" - what's not to like? :)

**Comment by bendk** [comment]
It makes a lot of sense to me as well and also reminds me of the Firefox JS work.  That ended up working out, so I think this effort could too.

So, would each call to Rust would go like this: JS -> secondary scaffolding -> primary scaffolding -> actual rust function?

I wonder if you could skip the primary scaffolding altogether.  Most of the existing complexity is defining `FfiConverter` impls for the types.  The actual scaffolding calls are fairly easy to put together.  I'm not against going through the extra function, I'd just be interested to know what prevents this and if we could update `FfiConverter` to fix those issues.

> In the secondary scaffolding, use -> Result<(), WasmCallStatus> instead of &mut RustCallStatus. From my research wasm_bindgen does not provide built-in support for passing in arbitrary structs as &mut from JS. However, wasm_bindgen does provide good enough support for returning Result. On Rust side, the function simply returns a Result, where the error is a type convertable to JS ABI. On JS side, when an error is returned, intermediate JS bindings throw the error value. This intermediate error value can be caught, refined into a Uniffi error class, and rethrown.

This is a much nicer API IMO, and it would be nice if the other bindings did something similar.  From a quick read, it seems like `wasm-bindgen` supports `Result<T, JsValue>`.  Is there a way to represent a `RustBuffer` as a `JsValue`?

**Comment by arg0d** [comment]
@bendk you are right, after digging into Rust scaffolding, it looks like it would be better to simply provide an alternate scaffolding implementation for WASM. I will investigate this further.

Yes, `RustBuffer` can be converted to `JsValue`, but it needs to be annotated with `#[wasm_bindgen]`. I would really like to avoid this, because it would mean having to insert WASM code into `uniffi-rs` upstream. But I think this can be avoided by simply created an identical `WasmBuffer` struct, and WASM scaffolding could simply map between `RustBuffer` and `WasmBuffer` as needed.

Why do you think that callback interfaces are tricky? Rust can easily call JS by using [js_sys::JsFunction](https://rustwasm.github.io/wasm-bindgen/api/js_sys/struct.Function.html). 

It looks like my department is deprioritizing WASM JS bindings for this quarter, but I have hope to come back to this within the next few months.

**Comment by bendk** [comment]
Makes sense to me, excited to see how this progresses.

I don't think there's anything fundamentally hard about callback interfaces, it's just felt like they're a lot of work to implement relative to the rest of the features.

**Comment by badboy** [comment]
> what about using wasmer? [wasmer.io](https://wasmer.io)

What about it? wasmer is a Wasm runtime. It can run any wasm you give it.
wasm support in UniFFI would mean you can build it in a way that other parts that can interact with wasm can interact with the UniFFI-powered parts.

**Comment by gyzerok** [comment]
Hello @arg0d! Have you had a chance to come back to this? I'd be highly interested.

**Comment by arg0d** [comment]
Hey @gyzerok, wasm bindings were deprioritized completely by the library team :( Due to small size of their library, and due to wasm limitations put on Rust code, it will be easier for them to maintain a separate JS version of the library.

**Comment by gyzerok** [comment]
I see thank you!

@bendk in earlier discussion it is mentioned that Mozilla seems to have some `uniffi` implementation for WASM used for Firefox desktop. Is there any chance you'd be willing to share those?

**Comment by badboy** [comment]
> I see thank you!
> 
> @bendk in earlier discussion it is mentioned that Mozilla seems to have some `uniffi` implementation for WASM used for Firefox desktop. Is there any chance you'd be willing to share those?

(not ben) we don't have any Wasm implementation. There's a custom JS integration in Gecko, but that's very Gecko-specific.

**Comment by gogo2464** [comment]
@arg0d as an user I tested wasm-bindgen and I would like to have uniffi-bindgen module with javascript directly inlcuded in order to avoid to keep my single udl file as required code modification for bindings.

**Comment by Skaldebane** [comment]
This would definitely be super useful. I'm trying to migrate an Android app to Kotlin Multiplatform, including Wasm support, but I have a Rust shared library. Using UniFFI is pretty great so far, It's just the WebAssembly integration that's missing.

**Comment by paxbun** [comment]
I'm seeking methods to implement WASM support for KMP. I'm using UniFFI 0.28.3, and `cargo build --target wasm32-unknown-unknown` worked well with Rust 1.81. However, I couldn't do anything with the resulting wasm file because there was no way to manipulate the stack pointer (which is implemented as the 0th global) from the JS (or Kotlin) side, and all UniFFI function needs a pointer to a caller-allocated RustCallStatus.

wasm-bindgen implements this by [generating](https://github.com/rustwasm/wasm-bindgen/blob/c35cc9369d5e0dc418986f7811a0dd702fb33ef9/crates/cli-support/src/js/mod.rs#L4339) a function called __wbindgen_add_to_stack_pointer. This function simply does:

```c
void* __wbindgen_add_to_stack_pointer(int offset) {
  return global_0 /* the stack pointer */ += offset;
}
```

If such a function is exposed by the resulting wasm, we can do:

```javascript
let status = wasm.exports.__wbindgen_add_to_stack_pointer(- /* sizeof RustCallStatus */);
wasm.exports.uniffi_some_function(arg1, arg2, ..., status);
```

Even if we use wasm-bindgen, we have another issue: the global stack pointer function generation function is not a public API. We have to somehow re-implement this on our own.

**Comment by joe-p** [comment]
For anyone wanting WASM support *alongside* UniFFI I've had a pretty good experience combining UniFFI with `wasm-bindgen` and `Tsify`. I created some [proc macros]( https://github.com/algorandfoundation/Algorand-Rust-FFIs/blob/master/crates/ffi_macros/src/lib.rs.) to make it easier to use both. In the future this might be a seperate crate we publish, but for now it has some implementation specifics. None the less, could be a good starting point for someone looking to target both UniFFI bindings and WASM

The one thing to note is that I have not experimented much with Objects, just functions and Records.
